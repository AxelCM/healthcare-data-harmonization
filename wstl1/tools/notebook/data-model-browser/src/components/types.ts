// Copyright 2020 Google LLC.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

/**
 * Meta data describing a top level entity/resource within a data model.
 */
export interface TopEntityMeta {
  /**
   * The name of the entity (e.g. Patient)
   */
  name: string;

  /**
   * A detailed description of the entity.
   */
  description: string;
}

/**
 * Meta data describing all sub entities/fields within a data model.
 */
export interface SubEntityMeta {
  /**
   * Name of the sub entity/field.
   */
  name: string;

  /**
   * A flag indicating whether the field is required.
   */
  required: boolean;

  /**
   * The type of the sub entity. The type could either be a primitive in the
   * data model (e.g. string, number), or a reference to another complex type
   * (e.g. Date, Identifier in FHIR).
   */
  type: string;

  /**
   * A detailed description of the sub entity or field.
   */
  description: string;

  /**
   * The JSON schema definition of the sub entity.
   */
  // tslint:disable-next-line:no-any
  schema?: any;

  /**
   * A flag indicating whether the field contains a complex type which can be
   * further explored or a primitive type that can not be further explored.
   */
  clickable: boolean;
}

/**
 * A result of a search query.
 */
export interface SearchResult {
  /**
   * The path of the search result, including the associated data model.
   * For example, FHIR > Patient > address.
   */
  path: string;

  /**
   * The name of the top level resource associated with the search result.
   */
  resource: string;

  /**
   * The field associated with the search result. If the selected search result
   * was for a top level entity, then the field will be empty.
   */
  field: string;
}
