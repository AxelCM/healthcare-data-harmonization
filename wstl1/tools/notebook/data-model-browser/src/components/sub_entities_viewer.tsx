// Copyright 2020 Google LLC.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// tslint:disable-next-line:enforce-name-casing
import * as React from 'react';

import {makeStyles} from '@material-ui/core/styles';
import ClickAwayListener from '@material-ui/core/ClickAwayListener';
import Table from '@material-ui/core/Table';
import TableHead from '@material-ui/core/TableHead';
import TableBody from '@material-ui/core/TableBody';
import TableCell from '@material-ui/core/TableCell';
import TableRow from '@material-ui/core/TableRow';

import {InspectTooltip} from './inspect_tool_tip';
import {SubEntityTooltip} from './sub_entity_tool_tip';
import {SubEntityMeta} from './types';

const localStyles = makeStyles({
  span: {
    cursor: 'pointer',
    '&:hover': {
      textDecoration: 'underline',
    },
  },
});

/**
 * Properties to customize the SubEntitiesViewer component.
 */
interface Props {
  /**
   * A list of all of the sub entites/fields renderable by the component.
   */
  subEntities: SubEntityMeta[];

  /**
   * The name of the selected/active sub level entity.
   */
  selected: string;

  /**
   * Callback invoked when a specific top level entity is inspected.
   */
  onInspect: (name: string) => void;

  /**
   * Callback invoked when a specific sub level entity is selected.
   */
  onSelect: (name: string) => void;
}

/**
 * Component to visualize all the sub entites/fields within a data model.
 */
// tslint:disable-next-line:enforce-name-casing React component.
export function SubEntitiesViewer(props: Props) {
  const [open, setOpen] = React.useState(false);
  const handleTooltipClose = () => {
    setOpen(false);
  };
  const classes = localStyles();
  let keyIndex = 0;
  return (
    <ClickAwayListener onClickAway={handleTooltipClose}>
      <Table>
        <TableHead>
          <TableRow>
            <TableCell>Field name</TableCell>
            <TableCell>Type</TableCell>
          </TableRow>
        </TableHead>
        <TableBody>
          {props.subEntities.map(meta => {
            const isSelected = meta.name === props.selected;
            return (
              <InspectTooltip
                key={keyIndex++}
                title={SubEntityTooltip(meta)}
                open={isSelected && open}
                disableFocusListener
                disableHoverListener
                disableTouchListener
              >
                <TableRow
                  hover
                  key={meta.name}
                  onClick={e => {
                    setOpen(true);
                    props.onInspect(meta.name);
                  }}
                  selected={isSelected}
                >
                  <TableCell component="th" scope="row">
                    <span
                      className={classes.span}
                      onClick={e => {
                        if (meta.clickable) {
                          props.onSelect(meta.name);
                        }
                      }}
                    >
                      {meta.name}
                    </span>
                  </TableCell>
                  <TableCell>
                    <span>{meta.type}</span>
                  </TableCell>
                </TableRow>
              </InspectTooltip>
            );
          })}
        </TableBody>
      </Table>
    </ClickAwayListener>
  );
}
