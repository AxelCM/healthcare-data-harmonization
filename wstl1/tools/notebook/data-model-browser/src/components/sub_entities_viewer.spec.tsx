// Copyright 2020 Google LLC.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

// tslint:disable-next-line:enforce-name-casing
import * as React from 'react';
import {shallow} from 'enzyme';
import {SubEntitiesViewer} from './sub_entities_viewer';
import {SubEntityMeta} from './types';

describe('SubEntitiesViewer', () => {
  it('should render correctly', () => {
    const entities: SubEntityMeta[] = [
      {
        name: 'Test1',
        required: true,
        type: 'string',
        description: 'Verbose details',
        clickable: true,
      }
    ];

    const selected = 'Test1';
    const onInspect = jest.fn();
    const onSelect = jest.fn();
    const wrapper = shallow(
      <SubEntitiesViewer
        subEntities={entities}
        onInspect={onInspect}
        onSelect={onSelect}
        selected={selected}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });
});
