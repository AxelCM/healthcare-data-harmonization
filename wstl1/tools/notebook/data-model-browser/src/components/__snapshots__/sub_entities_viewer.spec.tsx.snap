// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SubEntitiesViewer should render correctly 1`] = `
<WithStyles(ForwardRef(Table))>
  <WithStyles(ForwardRef(TableHead))>
    <WithStyles(ForwardRef(TableRow))>
      <WithStyles(ForwardRef(TableCell))>
        Field name
      </WithStyles(ForwardRef(TableCell))>
      <WithStyles(ForwardRef(TableCell))>
        Type
      </WithStyles(ForwardRef(TableCell))>
    </WithStyles(ForwardRef(TableRow))>
  </WithStyles(ForwardRef(TableHead))>
  <WithStyles(ForwardRef(TableBody))>
    <InspectTooltip
      disableFocusListener={true}
      disableHoverListener={true}
      disableTouchListener={true}
      key="0"
      open={true}
      title={
        <React.Fragment>
          <div>
            Field:
            <p>
              Test1
            </p>
          </div>
          <ForwardRef(WithStyles) />
          <div>
            Required:
            <p>
              yes
            </p>
          </div>
          <div>
            Type:
            <p>
              string
            </p>
          </div>
          <div>
            Scope and Usage:
          </div>
          <p>
            Verbose details
          </p>
        </React.Fragment>
      }
    >
      <WithStyles(ForwardRef(TableRow))
        hover={true}
        key="Test1"
        onClick={[Function]}
        selected={true}
      >
        <WithStyles(ForwardRef(TableCell))
          component="th"
          scope="row"
        >
          <span
            className="makeStyles-span-1"
            onClick={[Function]}
          >
            Test1
          </span>
        </WithStyles(ForwardRef(TableCell))>
        <WithStyles(ForwardRef(TableCell))>
          <span>
            string
          </span>
        </WithStyles(ForwardRef(TableCell))>
      </WithStyles(ForwardRef(TableRow))>
    </InspectTooltip>
  </WithStyles(ForwardRef(TableBody))>
</WithStyles(ForwardRef(Table))>
`;
